### DISCUSSION TITLE
toHaveBeenCalled() inconsistency with jest - why require spy and any workaround?

### DISCUSSION THREAD

**Discussion by brianbattenfeld** [vitest-dev/vitest#5850] [Q&A]
I'm experimenting with vitest at the moment and noticing what feels like an odd discrepency between vitest and jest's implementation of toHaveBeenCalled().

In a scenario where I simply want to pass a mock function like a form submit handler for instance, I would expect to be able to define the function, pass it in as a prop (working with react) and write `expect(functionName).toHaveBeenCalled()`

This wasn't working and I noticed that the [vitest docs](https://vitest.dev/api/expect.html#tohavebeencalled) suggest that a spy function is required — which requires that the function we're testing be a method on a class. 

I'm mostly just curious whether I'm misunderstanding something and if not, why implement this so differently than jest? Requiring a spy function, and requiring our functions to be defined within a class just feels convoluted.

Really curious to hear thoughts or an explanation though.

**Comment by hi-ogawa** [comment]
Can you provide a full code to show what's not working (and what's working on Jest)? When Vitest says "spy function", that should include `vi.fn` https://vitest.dev/api/vi.html#vi-fn, so you should be able to write something like this, for example:

```tsx
test("...", () => {
  const myFn = vi.fn();
  render(<button onClick={myFn} />);
  fireSomeClickEvent(...);
  expect(myFn).toHaveBeenCalled();
})
```

**Reply by brianbattenfeld** [reply]
This is what I was expecting to have worked. I'll take another look today and post a sample if I'm still having issues.

**Reply by brianbattenfeld** [reply]
So I've got pretty much the same thing here. For reference, there's 1 single button (the submit) rendered within this component and the onSubmit prop is valid etc.

I get the following in my console though:
`AssertionError: expected "spy" to be called at least once`

```javascript
test("calls the onSubmit function when the form is submitted with valid data", () => {
    const onMockSubmit = vi.fn();
    render(<FormComponent  onSubmit={onMockSubmit} />);
    fireEvent.click(screen.getByRole("button"));
    expect(onMockSubmit).toHaveBeenCalled();
  });
```

**Reply by sheremet-va** [reply]
It's probably resolved async, try to await the click call
